Add tutorial on physics addition (#956)
NVIDIA/physicsnemo: docs/tutorials/physics_addition.rst

> Adding Physics-based constraints using PhysicsNeMo

Author: Great points. Added some opening text ad modified where applicable. Please have a look.

A full example using this loss can be found in the `Molecular Dynamics Example <../examples/molecular_dynamics/lennard_jones/README.rst>`_

Adding PDE constraints
~~~~~~~~~~~~~~~~~~~~~~~

Sometimes, the governing equations for the problem can include Partial Differential Equations.
This is especially true for problems in physics and other scientific domains. Computing these
PDEs on model predictions involves computing spatial and temporal gradients and the methods
used can vary based on the model architecture. 

PhysicsNeMo-Sym aims to simplify this process. 
The `PhysicsInformer <https://docs.nvidia.com/deeplearning/physicsnemo/physicsnemo-sym/api/physicsnemo.sym.eq.html#module-physicsnemo.sym.eq.phy_informer>`_
utility can compute the PDE losses on point clouds, grids or graphs/meshes using
techniques such as Automatic Differentiation, Finite Difference, Meshless Finite Difference,
Least Squares, Spectral Differentiation, etc. 

The ``PhysicsInformer`` class requires equations to compute as a parameter. The equation
must be in the form of PhysicsNeMo-Sym's PDE. Custom PDEs are also supported. 

Some general comments for the ``PhysicsInformer``:

> - Using PhysicsInformer, you can physics-inform almost any model architecture

Reviewer: ```suggestion
- Using `PhysicsInformer`, you can add physics loss terms to almost any model architecture
```

  (point-cloud based, grid based, graph based, etc.), e.g. MLPs, DeepONets, FNOs, CNNs, Diffusion Models,
  Graph networks, etc.
- The utility supports gradients via Automatic Differentiation, Spectral,

>   Finite Difference, Meshless Finite Difference and Least Squares methods.

Reviewer: Nevermind, seen below.
```suggestion
  Finite Difference, Meshless Finite Difference and Least Squares methods. (See the following section for more info.)
```

>   Finite Difference, Meshless Finite Difference and Least Squares methods.

Author: Great point. I also provided links to API docs for individual sections, just in case.

  predict field values on stencil points in addition to the original points. The points can come from a point cloud, 
  grid (structured or unstructured), but most suitable for point clouds. 
  Fast, but can present numerical instability if ``fd_dx`` is too small.

- **Finite Difference**: Numerical gradient method suitable for models that output predictions on
  structured grids with uniform spacing (each dimension can have a different spacing of its own).

- **Spectral Derivatives**: Numerical gradient method suitable for models that output predictions
  on structured grids with uniform spacing (each dimension can have a different spacing of its own) and have
  periodic boundaries.

- **Least Squares Method**: Numerical gradient method most suitable for models predict output on unstructured
  grids or structured grids with non-uniform spacing. 

Computing PDE losses using Automatic Differentiation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The code below shows an example of using the ``autodiff`` method to compute the residuals.
A few things to note when using the ``autodiff`` method:

- Ensure the model is differentiable, e.g. a model that uses ReLU activation

>   function will have it's second derivatives zero. 

Reviewer: Actually, on second thought:
* First-order PDEs need $C^1$-continuity
* Second-order PDEs (e.g., Navier-Stokes) need $C^2$-continuity.
* And so on...

The example activation functions we've given here are $C^\infty$-continuous, so this is fine, but this may be something to highlight to the user here.

>   function will have it's second derivatives zero. 

Reviewer: Related: we can recommend that users who want to use ReLU functions use numerical derivatives instead - but in general, this will produce poor results for second-order PDEs if you try to use physics loss terms, even if one uses numerical derivatives. Fundamentally, the function space needs to admit curvature in order for a solution of a second-order PDE to be valid (in the non-trivial case), so I worry we're steering users in the wrong direction here.

>   function will have it's second derivatives zero. 

Author: Great point. Added.

> - Set ``.requires_grad_()`` to ``True``. 

Reviewer: ```suggestion
- For all spatial coordinate tensors (e.g., `x`, `y`, and `z`), call the method ``x.requires_grad_(True)`` to enable gradient tracking. 
```

  PhysicsNeMo Sym allows you to load STL files and also define custom geometries
  using Constructive Solid Geometry and use it for computing the SDF. The 
  `geometry module documentation from PhysicsNeMo Sym <https://docs.nvidia.com/deeplearning/physicsnemo/physicsnemo-sym/user_guide/features/csg_and_tessellated_module.html#>`_
  provides a comprehensive documentation of this functionality. The code below shows
  a sample implementation of this

  .. code-block:: python

    import numpy as np
    from physicsnemo.sym.geometry.tessellation import Tessellation

    # read the Stanford Bunny stl
    geo = Tessellation.from_stl("./Stanford_Bunny.stl")

    # compute the SDF on the (0, 0, 0) points
    sdf = geo.sdf(
            {
                "x": np.array([[0]]),   # each coordinate must have shape (N, 1)
                "y": np.array([[0]]),
                "z": np.array([[0]]),
            },
        params={}
    )["sdf"]

A few examples using SDF during training / inference can be found in the 
`External Aerodynamics using DoMINO Example <../examples/cfd/external_aerodynamics/domino/README.rst>`_, 
`Datacenter CFD example <../examples/cfd/datacenter/README.rst>`_.

> Sampling point clouds

Reviewer: ```suggestion
Sampling Point Clouds
```
